pkg/controller/baremetalhost/baremetalhost_controller.go

```diff
@@ -198,54 +198,54 @@ func (r *ReconcileBareMetalHost) Reconcile(request reconcile.Request) (result re
+				host.Status.ErrorMessage = err.Error()
+				saveErr := r.saveStatus(host)
+				if saveErr != nil {
+					return reconcile.Result{Requeue: true}, saveErr
+				}
 				// Only publish the event if we do not have an error
 				// after saving so that we only publish one time.
 				r.publishEvent(request,
-					host.NewEvent("Discovered", "Discovered host without BMC address"))
+					host.NewEvent("Discovered", fmt.Sprintf("Discovered host with unusable BMC details: %s", err.Error())))
 			}
-			// Without the address we can't do any more so we return here
-			// without checking for an error.
-			return reconcile.Result{Requeue: true}, err
-		}
-		reqLogger.Info("nothing to do for discovered host without BMC address")
-		return reconcile.Result{}, nil
-	}
-	if host.Spec.BMC.CredentialsName == "" {
-		reqLogger.Info(bmc.MissingCredentialsMsg)
-		dirty := host.SetOperationalStatus(metalkubev1alpha1.OperationalStatusDiscovered)
-		if dirty {
-			err = r.saveStatus(host)
-			if err != nil {
-				// Only publish the event if we do not have an error
-				// after saving so that we only publish one time.
-				r.publishEvent(request,
-					host.NewEvent("Discovered", "Discovered host without BMC credentials"))
+			return reconcile.Result{Requeue: true, RequeueAfter: hostErrorRetryDelay}, nil
+		// In the event a credential secret is defined, but we cannot find it, or the secret is
+		// missing the required fields, or the BMC address is defined but malformed we set the
+		// host into an error state and requeue it for evaluation later
+		case *ResolveBMCSecretRefError, *bmc.CredentialsValidationError, *bmc.UnknownBMCTypeError:
+			saveErr := r.setErrorCondition(request, host, err.Error())
+			if saveErr != nil {
+				return reconcile.Result{Requeue: true}, saveErr
 			}
-			// Without any credentials we can't do any more so we return
-			// here without checking for an error.
-			return reconcile.Result{Requeue: true}, err
+			// Only publish the event if we do not have an error
+			// after saving so that we only publish one time.
+			r.publishEvent(request, host.NewEvent("BMCCredentialError", err.Error()))
+			return reconcile.Result{Requeue: true, RequeueAfter: hostErrorRetryDelay}, nil
+		default:
+			return reconcile.Result{}, errors.Wrap(err, "An unhandled failure occurred validating the BMC secret")
 		}
-		reqLogger.Info("nothing to do for discovered host without BMC credentials")
-		return reconcile.Result{}, nil
 	}
 
-	// Load the credentials for talking to the management controller.
-	bmcCreds, bmcCredsSecret, err := r.getValidBMCCredentials(request, host)
-	if err != nil {
-		return reconcile.Result{}, errors.Wrap(err, "BMC credentials are invalid")
-	}
-	if bmcCreds == nil {
-		// We do not have valid credentials, but did not encounter a
-		// retriable error in determining that. Reconciliation is
-		// complete until something about the secrets change.
-		return reconcile.Result{}, nil
+	// Make sure the secret has the correct owner.
```

Comment:
We need to do this step before reporting an error with the secret contents. It could happen down around line 736 where we load the secret object, or up after line 204 but before the error from that call is checked.

Reply:
Moved to right after we retrieve the secret.  It will return a SaveBMCCredentialsSecretOwnerError error (in the spirit of typing) but since we do not switch on it, we will return the error to the Reconciler loop which should immediately try again.  This seems appropriate given this type of error _should_ be transient baring any permissions issues.